### ISSUE TITLE
gRPC stream client OutOfDirectMemoryError, when send ton of data

### ISSUE THREAD

**Issue by wu-sheng** [grpc/grpc-java#2424]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
1.0.1

### What JVM are you using (`java -version`)?
1.8

### What did you do?
If possible, provide a recipe for reproducing the error.
Netty OutOfDirectMemoryError

### What did you expect to see?
send data as quickly as possible

### What did you see instead?
not to trigger OutOfDirectMemoryError

ref issue: #2247 

test case:
```java
for (int i = 0; i < 1000000; i++) {
            StreamObserver<RequestSpan> requestSpanStreamObserver =
                    spanStorageServiceStub.storageRequestSpan(new StreamObserver<SendResult>() {
                        @Override
                        public void onNext(SendResult sendResult) {

                        }

                        @Override
                        public void onError(Throwable throwable) {
                            throwable.printStackTrace();
                        }

                        @Override
                        public void onCompleted() {
                            endTime2 = System.currentTimeMillis();
                        }
                    });
            for (int j = 0; j < 10; j++) {
                requestSpanStreamObserver.onNext(requestSpan);
            }

            ClientCallStreamObserver<RequestSpan> newRequestSpanStreamObserver =
                    (ClientCallStreamObserver<RequestSpan>) requestSpanStreamObserver;

            while (!newRequestSpanStreamObserver.isReady()) {
                Thread.sleep(1);
            }

            ackSpanStreamObserver.onCompleted();
            requestSpanStreamObserver.onCompleted();
```

This single thread code send ton of data to server,  casue OutOfDirectMemoryError. more discussion detail, see the ref issue #2247

**Comment by carl-mastrangelo** [comment]
You need to respect the flow control pushback.   Basically, before calling `onNext()`, you should check `isReady()`.  If it is false, you can cast `requestSpanStreamObserver` to a `ClientCallStreamObserver`, and add an onReady handler with `setOnReadyHandler`

**Comment by ejona86** [comment]
@carl-mastrangelo, you can look at some of the discussion at end of #2247. I agree that the usage of `isReady()` isn't right, but it isn't wrong enough to cause the OOM. I think the issue is more of a lack/incomplete flow control for new RPCs, and it will take some wider discussion.

Short-term solutions is for the application to limit the number of concurrent RPCs is does. This is pretty common, and probably why we haven't heard about any such issue earlier. Even if we improve the flow control, the application will probably still need to limit the number of RPCs. If the server is configured with `NettyServerBuilder.maxConcurrentCallsPerConnection()`, then it will probably "fix" the isReady() behavior.

**Comment by carl-mastrangelo** [comment]
It is wrong enough.  It doesn't check if the call is ready before sending.  That's the line in the sand that we draw.  Check the FlowControl signals, or be subject to OOM.

**Comment by ejona86** [comment]
@carl-mastrangelo, I don't disagree, but that only makes it off-by-one RPC. So even if the readiness handling was fixed, the problem experienced would remain.

**Comment by carl-mastrangelo** [comment]
Actually, I seem to recall us always returning ready as long as the stream is allocated and less then DEFAULT_ONREADY_THRESHOLD   has been sent.  That means as long as the stream starts it is subject to oom.

**Comment by ejona86** [comment]
We do actually wait until the headers are written, but that doesn't matter much since headers are no longer put into the scheduler.

https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientStream.java#L221
https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientHandler.java#L387

(there could be other things that still delay isReady/onReady, but I think stream allocation is the primary one)

**Comment by wu-sheng** [comment]
@ejona86 ,
Maybe right now, I should set `NettyServerBuilder.maxConcurrentCallsPerConnection()` on my server side, and run my test again? Is it the right way for now?

And waiting your new feature to release?

**Comment by ejona86** [comment]
So I talked it over, and it seems we're still unsure what to do for Java. I'd suggest limiting your code to `x` (some hard-coded constant) concurrent RPCs, like with a semaphore. You could alternatively use `NettyServerBuilder.maxConcurrentCallsPerConnection()` on the server, but I'm less excited about that because it is really a client-side issue and would behave similarly to the semaphore.

C core has many more options than we do in Java, because of the difference in threading models. C's solution would _send_ no more than `x` RPCs concurrently; after a message was sent you could start another (so no dependence on how long the server takes, other than writing to the network). This doesn't work well in Java via `isReady()` because it would add too much latency to `isReady()` in the common case. Java could have a callback for when the message has been consumed, but there is a cost to each callback and this one would be rarely used. Maybe we could have an option (probably a setter on `*Call`, and `*CallStreamObserver`) so the callback could be enabled only when it will be used.

**Comment by wu-sheng** [comment]
I think this option is good enough for me. Looking for your release.

**Comment by olivere** [comment]
We're seeing the same type of error with a Go server and a Java client. Has anything happened in the recent releases to prevent this on the Java client side, or make it easier to work around?

**Comment by larry-safran** [comment]
@ejona86 doesn't `maxConcurrentRequests` now address this?

**Comment by ejona86** [comment]
@larry-safran, no, nothing has changed here. `maxConcurrentCallsPerConnection()` can tweak the onReady() behavior for a very specific circumstance, but it really doesn't protect the client from "too many concurrent RPCs." This same problem exists for unary and wouldn't be solved with stream-level flow control there.

I don't think there's anything to do here, though. The application just needs to limit itself, just like it does when creating Threads and other similar resources.
